### PR TITLE
fluids: Add external SGS DD inference capability

### LINKED ISSUE #1326: Fluids: Allow for external DD inference for SGS model
Instead of only having the internal neural-network calculation for the SGS data-driven model, allow for external libraries to be used for the inference, such as PyTorch.

Current code process:
- SGS Data-driven model is all setup in `problems/sgs_dd_model.c`
    - This creates `sgs_dd_data->op_nodal_evaluation_ctx` to calculate the SGS stresses at nodes and `sgs_dd_data->op_sgs_apply_ctx` to calculate the contribution of the SGS stress to the RHS residual
- During each call to `IFunction`, `SgsDDModelApplyIFunction` is called to add the contribution of the SGS stress

New (proposed) code process:
- Have `SgsDDModelApplyIFunction` call a separate function (via function pointer) to calculate the SGS stress at nodes
    - Function pointer should have the signature `(User user, Vec Q_loc, Vec VelocityGradient, Vec SGSNodal_loc)`
- `SgsDDModelSetup` should set that function pointer based on what should be used.

### PR DESCRIPTION
**Pull request by jrwrigh**
This adds the capability to use external libraries to do the inference of the SGS DD model. Currently, this is just a framework with the native libCEED inference method enabled. External libraries will be added in later.

More verbosely, this introduces to different ways to evaluate the data-driven SGS model: fused and sequential.

Fused is what was implemented before, where the input creation, model inference, and output handling were all done in a single CeedOperator. This requires a native libCEED inference implementation.

Sequential has separate function calls/CeedOperators for input creation, model inference, and output handling. This allows for functions calling external libraries to be used for the model inference step.

Closes #1326

### CODE REVIEW

File: examples/fluids/problems/sgs_dd_model.c [hunk 1]
```diff
@@ -14,24 +14,26 @@
 typedef struct {
   CeedElemRestriction  elem_restr_grid_aniso, elem_restr_sgs;
   CeedVector           grid_aniso_ceed;
-  CeedQFunctionContext sgsdd_qfctx;
-} *SgsDDModelSetupData;
+  CeedQFunctionContext sgsdd_qfctx, ifunction_qfctx;
```
Comment (jeremylt) [inline examples/fluids/problems/sgs_dd_model.c:17]: Tangentially related - if you have access to an operator, you have access to the QFunction and QFunctionContext. How much can you trim back these structs throughout the fluids code?

Comment (jrwrigh) [reply examples/fluids/problems/sgs_dd_model.c:17]: I'm assuming this is more about `ifunction_qfctx` than `sgsdd_qfctx`. 

Well, the main complication is that the IFunction operator is a composite operator and not all the suboperators have the same qfunction context. I'm not sure if there's a way to grab a suboperator by name (though that could be easily coded up  for the library or for the fluid's example exclusively I suppose).

Note also that this struct lives only inside this file; it's primary job is to ease the verbosity of function parameters within this setup function.

Comment (jeremylt) [reply examples/fluids/problems/sgs_dd_model.c:17]: Ah, I got around this by tracking the index of the relevant suboperator. Overall I've been on a campaign to reduce how much data we drag around in structs in Ratel but haven't started waging war on structs here in the libCEED repo

**Comment by jeremylt** [comment]
Is this still planning on being merged? Looks reasonable to me

**Comment by jrwrigh** [comment]
Yep, it's been ready for review for a couple months, but also isn't pressing. (though it'd be nice to go ahead and have it merged in obviously).

File: examples/fluids/problems/sgs_dd_model.c [hunk 2]
```diff
@@ -150,8 +164,259 @@ PetscErrorCode SgsDDModelSetupNodalEvaluation(Ceed ceed, User user, CeedData cee
...
+    }
+
+    PetscCall(DMClone(sgs_dd_data->dm_sgs, &sgs_dd_data->dm_dd_outputs));
+    PetscCall(PetscObjectSetName((PetscObject)sgs_dd_data->dm_dd_outputs, "Data-Driven Model Outputs"));
+    PetscCall(DMSetupByOrder_FEM(PETSC_TRUE, PETSC_TRUE, degree, 1, q_extra, 1, &sgs_dd_data->num_comp_outputs, sgs_dd_data->dm_dd_outputs));
+    PetscCall(DMGetLocalSection(sgs_dd_data->dm_dd_outputs, &section));
+    PetscCall(PetscSectionSetFieldName(section, 0, ""));
+    for (CeedInt i = 0; i < sgs_dd_data->num_comp_outputs; i++) {
+      char component_name[PETSC_MAX_PATH_LEN];
+      PetscCall(PetscSNPrintf(component_name, sizeof component_name, "DataDrivenOutput%" CeedInt_FMT, i + 1));
```
Comment (jeremylt) [inline examples/fluids/problems/sgs_dd_model.c:277]:
```suggestion
    for (CeedInt i = 0; i < sgs_dd_data->num_comp_inputs; i++) {
      char component_name[PETSC_MAX_PATH_LEN];
      
      PetscCall(PetscSNPrintf(component_name, sizeof component_name, "DataDrivenInput%" CeedInt_FMT, i + 1));
      PetscCall(PetscSectionSetComponentName(section, 0, i, component_name));
    }

    PetscCall(DMClone(sgs_dd_data->dm_sgs, &sgs_dd_data->dm_dd_outputs));
    PetscCall(PetscObjectSetName((PetscObject)sgs_dd_data->dm_dd_outputs, "Data-Driven Model Outputs"));
    PetscCall(DMSetupByOrder_FEM(PETSC_TRUE, PETSC_TRUE, degree, 1, q_extra, 1, &sgs_dd_data->num_comp_outputs, sgs_dd_data->dm_dd_outputs));
    PetscCall(DMGetLocalSection(sgs_dd_data->dm_dd_outputs, &section));
    PetscCall(PetscSectionSetFieldName(section, 0, ""));
    for (CeedInt i = 0; i < sgs_dd_data->num_comp_outputs; i++) {
      char component_name[PETSC_MAX_PATH_LEN];
      
      PetscCall(PetscSNPrintf(component_name, sizeof component_name, "DataDrivenOutput%" CeedInt_FMT, i + 1));
```
minor - usually we have newlines after variable declaration blocks

**Review by jeremylt — APPROVED** [review]

**Comment by jedbrown** [comment]
Thanks @jrwrigh for developing this feature and @jeremylt for reviewing.

**Comment by jrwrigh** [comment]
Welcome!

**Pull request merged**